### PR TITLE
ARROW-17405: [Doc][Java] C Data Interface library able to compile with mvn command

### PR DESCRIPTION
**Pull request by davisusanibar**
C Data Interface library (.so / .dylib) able to compile with mvn command:

$ mvn clean generate-resources -Pgenerate-cdata-dylib_so -N

### CODE REVIEW

File: docs/source/developers/java/building.rst [hunk 1]
```diff
@@ -47,40 +47,78 @@ repository:
...
 We can build these manually or we can use `Archery`_ to build them using a Docker container
 (This will require installing Docker, Docker Compose, and Archery).
 
+Note: If you are building on Apple Silicon, be sure to use a JDK version that was compiled
+for that architecture. See, for example, the `Azul JDK <https://www.azul.com/downloads/?os=macos&architecture=arm-64-bit&package=jdk>`_.
+
+Maven
+~~~~~
+
+- To build only the C Data Interface library:
```
Comment (kou) [inline docs/source/developers/java/building.rst:103]: If we itemize this subsection, we need to indent the following `.. code-block::` to put into this subsection.

Comment (davisusanibar) [reply docs/source/developers/java/building.rst:103]: Changed

**Comment by lwhite1** [comment]
These changes look ok to me. Is the Java Windows failure random/intermittent or is there something that needed fixing there?

**Comment by kou** [comment]
Could you rebase on the master?

**Comment by davisusanibar** [comment]
> Could you rebase on the master?

Merged with master

File: docs/source/developers/java/building.rst [hunk 2]
```diff
@@ -47,25 +47,156 @@ repository:
...
+        $ cd arrow
+        $ mkdir -p java-dist java-jni
+        $ cmake \
+            -S java \
+            -B java-jni \
+            -DCMAKE_BUILD_TYPE=Release \
+            -DCMAKE_INSTALL_PREFIX=java-dist/lib \
+            -DARROW_JAVA_JNI_ENABLE_DEFAULT=OFF \
+            -DARROW_JAVA_JNI_ENABLE_C=ON \
+            -DBUILD_TESTING=OFF
```
Comment (kou) [inline docs/source/developers/java/building.rst:134]: Could you sort this list in alphabetical order?

File: docs/source/developers/java/building.rst [hunk 3]
```diff
@@ -47,25 +47,156 @@ repository:
...
+            -DARROW_PARQUET=ON \
+            -DARROW_PLASMA=ON \
+            -DARROW_PLASMA_JAVA_CLIENT=ON \
+            -DARROW_S3=ON \
+            -DARROW_USE_CCACHE=ON \
+            -DCMAKE_BUILD_TYPE=Release \
+            -DCMAKE_INSTALL_LIBDIR=lib \
+            -DCMAKE_INSTALL_PREFIX=java-dist \
+            -DCMAKE_UNITY_BUILD=ON \
+            -DARROW_DEPENDENCY_SOURCE=BUNDLED
```
Comment (kou) [inline docs/source/developers/java/building.rst:172]: Could you sort this list in alphabetical order?

File: docs/source/developers/java/building.rst [hunk 4]
```diff
@@ -47,25 +47,156 @@ repository:
...
+        $ mkdir -p java-dist java-jni
+        $ cmake \
+            -S java \
+            -B java-jni \
+            -DCMAKE_BUILD_TYPE=Release \
+            -DCMAKE_INSTALL_PREFIX=java-dist/lib \
+            -DCMAKE_PREFIX_PATH=$PWD/java-dist \
+            -DARROW_JAVA_JNI_ENABLE_DEFAULT=OFF \
+            -DARROW_JAVA_JNI_ENABLE_DATASET=ON \
+            -DBUILD_TESTING=OFF
```
Comment (kou) [inline docs/source/developers/java/building.rst:192]: Could you sort this list in alphabetical order?

File: java/pom.xml [hunk 5]
```diff
@@ -905,6 +905,70 @@
...
+                <configuration>
+                  <executable>cmake</executable>
+                  <commandlineArgs>
+                    -S java
+                    -B java-jni
+                    -DCMAKE_BUILD_TYPE=Release
+                    -DCMAKE_INSTALL_PREFIX=${arrow.c.jni.dist.dir}
+                    -DARROW_JAVA_JNI_ENABLE_DEFAULT=OFF
+                    -DARROW_JAVA_JNI_ENABLE_C=ON
+                    -DBUILD_TESTING=OFF
```
Comment (kou) [inline java/pom.xml:947]: Could you sort this list in alphabetical order?

**Review by kou — APPROVED** [review]
+1

We can merge this after we sort lists.

File: docs/source/developers/java/building.rst (hunk 2 above)
Comment (davisusanibar) [reply docs/source/developers/java/building.rst]: Sorted

File: docs/source/developers/java/building.rst (hunk 3 above)
Comment (davisusanibar) [reply docs/source/developers/java/building.rst]: Sorted

File: docs/source/developers/java/building.rst (hunk 4 above)
Comment (davisusanibar) [reply docs/source/developers/java/building.rst]: Sorted

**Pull request merged**